pkg/fuzzer/queue/queue.go

```diff
@@ -0,0 +1,375 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package queue
+
+import (
+	"context"
+	"math/rand"
+	"sync"
+
+	"github.com/google/syzkaller/pkg/hash"
+	"github.com/google/syzkaller/pkg/ipc"
+	"github.com/google/syzkaller/pkg/signal"
+	"github.com/google/syzkaller/pkg/stats"
+	"github.com/google/syzkaller/prog"
+)
+
+type Request struct {
+	Prog       *prog.Prog
+	NeedSignal SignalType
+	NeedCover  bool
+	NeedHints  bool
+	// TODO: passing ExecOpts directly is actually quite hacky -- this is a high-level interface
+	// and it should specify what is needed in high-level terms.
+	ExecOpts *ipc.ExecOpts
+
+	// If specified, the resulting signal for call SignalFilterCall
+	// will include subset of it even if it's not new.
+	SignalFilter     signal.Signal
+	SignalFilterCall int
+
+	// This stat will be incremented on request completion.
+	Stat *stats.Val
+
+	// Options needed by runtest.
+	BinaryFile   string // If set, it's executed instead of Prog.
+	Repeat       int    // Repeats in addition to the first run.
+	ReturnOutput bool
+
+	// The callback will be called on request completion in the LIFO order.
+	// If it returns false, all further processing will be stopped.
+	// It allows wrappers to intercept Done() requests.
+	callback DoneCallback
+
+	mu          sync.Mutex
+	initialized bool
+	result      *Result
+	done        chan struct{}
+}
+
+type DoneCallback func(*Request, *Result) bool
+
+func (r *Request) OnDone(cb DoneCallback) {
+	oldCallback := r.callback
+	r.callback = func(req *Request, res *Result) bool {
+		r.callback = oldCallback
```

Comment:
Is it better to restore the old one, or keep the current one?

I think the only affected party would be retryer, and it could:
```
func (r *retryer) Next() *Request {
	req := r.pq.tryNext()
	if req != nil {
		return req
	}
	req := r.base.Next()
	req.OnDone(r.done)
	return req
}
```

Fuzzer also installs own callbacks, but there it's moot. It won't reinstall the callback again, but also assumes processing won't be aborted after it.

Reply:
If we don't restore the old callback, we'd just keep calling the last one over and over again. Or do you mean restoring specifically in case of the callback returning false?

Overall, these callbacks are meant to form a kind of a stack, where the callbacks that were added earlier should not worry about further callbacks at all.

Reply:
Yes, I mean specifically the case when a callback returns false.
Without `r.callback = oldCallback` line we won't call the same callback, we will call the `oldCallback`, and we also get LIFO. 
The `r.callback = oldCallback` line only has very special effect on the `return false` case -- will all stages need to re-add itself again or not. For example, if an earlier callback already did `return true`, it won't be able to re-add itself.

Reply:
> it won't be able to re-add itself.

The callback is not supposed to add itself, the callbacks are to be set in `Next()`. 
```
type intermediateLayer struct {
  base queue.Source
}

func (layer *intermediateLayer) Next() *Request {
  req := layer.base.Next()
  req.OnDone(layer.onDone)
  return req
}

func (layer *intermediateLayer) onDone(req *Request, res *Result) bool {
  // (*)
}

```

Say a `Result` has gone up some callbacks (they returned true), then it hit some layer that decided that it better not propagate further. There are now two alternatives:
1) It will be resent. In which case `intermediateLayer.Next()` will have to return it at some point. Then all further layers on the `Next()` chain (the one that calls `intermediateLayer.Next()`, etc.) will set their callbacks again, if they determine that it's necessary.
2) it will release the result at some later time. In this case, `intermediateLayer` would just call `Done()` again. All layers that go on top of `intermediateLayer` don't need to care about this.

In both cases it seems fine for `OnDone()` to only set one-time callbacks. Or am I missing something?